components/ide/jetbrains/backend-plugin/src/main/kotlin/io/gitpod/jetbrains/remote/GitpodTerminalService.kt

```diff
@@ -0,0 +1,382 @@
+
+                        thisLogger()
+                                .error(
+                                        "Got an error while writing to '${supervisorTerminal.title}' terminal.",
+                                        throwable
+                                )
+                    }
+
+                    override fun onCompleted() {
+                        debug("'${supervisorTerminal.title}' terminal finished writing stream.")
+                        isWaitingForWritingResponse = false
+                    }
+                }
+
+        val writeTerminalRequestBuilder =
+                TerminalOuterClass.WriteTerminalRequest.newBuilder()
+                        .setAlias(supervisorTerminal.alias)
+
+        @Suppress("EXPERIMENTAL_IS_NOT_ENABLED")
+        @OptIn(DelicateCoroutinesApi::class)
+        val watchTerminalInputJob =
+                GlobalScope.launch {
+                    withContext(Dispatchers.IO) {
+                        while (isActive) {
+                            if (dataReceivedFromSupervisor.size() > 0) {
+                                val bytes = dataReceivedFromSupervisor.toByteArray()
+                                dataReceivedFromSupervisor.reset()
+                                terminalOutputWriter.write(bytes)
+                                terminalOutputWriter.flush()
+                            }
+
+                            val bytesAvailableOnTerminalInput = terminalInputReader.available()
+
+                            if (!isWaitingForWritingResponse && bytesAvailableOnTerminalInput > 0) {
+                                isWaitingForWritingResponse = true
+
+                                val bytesFromTerminalInput =
+                                        ByteArray(bytesAvailableOnTerminalInput)
+
+                                terminalInputReader.read(
+                                        bytesFromTerminalInput,
+                                        0,
+                                        bytesAvailableOnTerminalInput
+                                )
+
+                                val supervisorTerminalStdin =
+                                        ByteString.copyFrom(bytesFromTerminalInput)
+                                val writeTerminalRequest =
+                                        writeTerminalRequestBuilder
+                                                .setStdin(supervisorTerminalStdin)
+                                                .build()
+
+                                terminalServiceStub.write(
+                                        writeTerminalRequest,
+                                        writeTerminalResponseObserver
+                                )
+                            }
+                        }
+                    }
+                }
+
+        shellTerminalWidget.addListener {
+            debug("Terminal '${supervisorTerminal.title}' was closed via IDE.")
```

Comment:
It is only triggered when a user explicitly close the terminal, right? We should not call close if a user closes the entire thin client.

Reply:
Yeah, this should happen only when user clicks the "X" button on a specific terminal tab.
But let me confirm.